src/js/integrations/stalltracking.ts

```diff
@@ -0,0 +1,336 @@
+    if (!statsOnFinish) {
+      if (typeof endTimestamp !== "undefined") {
+        logger.log(
+          "[StallTracking] Stall measurements not added due to `endTimestamp` being set."
+        );
+      } else if (trimEnd) {
+        logger.log(
+          "[StallTracking] Stall measurements not added due to `trimEnd` being set but we could not determine the stall measurements at that time."
+        );
+      }
+
+      return null;
+    }
+
+    return {
+      stall_count: {
+        value: statsOnFinish.stall_count.value - statsOnStart.stall_count.value,
+      },
+      stall_total_time: {
+        value:
+          statsOnFinish.stall_total_time.value -
+          statsOnStart.stall_total_time.value,
+      },
+      stall_longest_time: statsOnFinish.stall_longest_time,
+    };
+  }
+
+  /**
+   * Logs the finish time of the span for use in `trimEnd: true` transactions.
+   */
+  private _logSpanFinish(
+    transaction: Transaction,
+    spanEndTimestamp: number
+  ): void {
+    if (
+      Math.abs(timestampInSeconds() - spanEndTimestamp) >
+      MARGIN_OF_ERROR_SECONDS
+    ) {
+      logger.log(
+        "[StallTracking] Span end not logged due to end timestamp being outside the margin of error from now."
+      );
+
+      const previousStats = this._statsAtTimestamp.get(transaction);
+
+      if (previousStats && previousStats.timestamp < spanEndTimestamp) {
+        // We also need to delete the stat for the last span, as the transaction would be trimmed to this span not the last one.
+        this._statsAtTimestamp.delete(transaction);
+      }
+    } else {
+      this._statsAtTimestamp.set(transaction, {
+        timestamp: spanEndTimestamp,
+        stats: this._getCurrentStats(transaction),
+      });
+    }
+  }
+
+  /**
+   * Get the current stats for a transaction at a given time.
+   */
+  private _getCurrentStats(transaction: Transaction): StallMeasurements {
+    return {
+      stall_count: { value: this._stallCount },
+      stall_total_time: { value: this._totalStallTime },
+      stall_longest_time: {
+        value: this._longestStallsByTransaction.get(transaction) ?? 0,
+      },
+    };
+  }
+
+  /**
+   * Start tracking stalls
+   */
+  private _startTracking(): void {
+    if (!this.isTracking) {
+      this.isTracking = true;
+      this._lastIntervalMs = Math.floor(timestampInSeconds() * 1000);
+
+      this._iteration();
+    }
+  }
+
+  /**
+   * Stops the stall tracking interval and calls reset().
+   */
+  private _stopTracking(): void {
+    this.isTracking = false;
+
+    if (this._timeout !== null) {
+      clearTimeout(this._timeout);
+      this._timeout = null;
+    }
+
+    this._reset();
+  }
+
+  /**
+   * Clears all the collected stats
+   */
+  private _reset(): void {
+    this._stallCount = 0;
+    this._totalStallTime = 0;
+    this._lastIntervalMs = 0;
+    this._longestStallsByTransaction = new Map();
+    this._runningTransactions = new Set();
+    this._statsAtTimestamp = new Map();
```

Comment:
let's try to avoid creating new objects.

```suggestion
    this._longestStallsByTransaction.clear();
    this._runningTransactions.clear();
    this._statsAtTimestamp.clear();
```